src/java/org/tensorics/core/tree/walking/Trees.java

```diff
@@ -197,16 +174,16 @@ public void onEvery(Node node) {
         return newRebuildingContext.getUpdatedOrSame(rootNode);
     }
 
-    /**
+	/**
      * walks through the tree, starting from the given rootNode and collects all the nodes which implement the given
      * class.
      * 
      * @param rootNode the node from which to start the search
      * @param nodeClassToFind the class of the nodes to find
      * @return a set of all found nodes, which implement the given class
      */
-    public static <T extends Node> Set<T> findNodesOfClass(Node rootNode, final Class<T> nodeClassToFind) {
-        final Set<T> foundNodes = new HashSet<>();
+    public static <T extends Node, C> Set<C> findNodesOfClass(T rootNode, final Class<C> nodeClassToFind) {

```

Comment:
I guess I'm missing something obvious ... but why shouldn't C (or T before) be restricted to 'C extends Node'?
The tree shall only contain and the walker only callback on Nodes I'd think. I'm wondering if this will ever return anything but an empty set if called with a Class<C> which does not extend Node.

If the empty set behaviour is needed, I'd at least document it clearly.